### PR TITLE
[WIP] Core doc update

### PR DESCRIPTION
**Pull request by rete**
BEGINRELEASENOTES
- Core documentation updates:
    - Signal/slot mechanism updated
    - Event interface updated
    - Plugin interface updated
- Added new sections in core documentation:
    - New section on StatusCode / exception and pre-processor macros
    - New section on XML parser features and `XMLParser` class
- Default theme set to `readthedocs`

ENDRELEASENOTES

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ pages:
   - User Manual:
     - Core tools: 'user-manual/core-tools.md'
     - The plugin system: 'user-manual/plugin-system.md'
-    - Event streaming: 'user-manual/event-streaming.md'
+    - Event interface: 'user-manual/event-interface.md'
```
Comment (apingault) [inline mkdocs.yml:9]: You should also update the reference in [plugin-system.md](https://github.com/DQM4HEP/dqm4hep-doc/pull/7/files#diff-697a7eedcca57fbb6391098afc0064ebR15)

File: docs/user-manual/core-tools.md [hunk 2]
```diff
@@ -143,9 +143,9 @@ int main() {
 
 By default, in most of the binaries provided by the DQM4hep packages, the main logger is configured to be a colored console.
 
-# Signal mechanism
+# Signal/slot mechanism
 
-The DQM4hep core library provide a class with a similar behavior as the Qt signal/slot mechanism. This is mainly used by the DQM4hep net library to disptach data received through the network to different callback functions, but can be re-used in many contexts. The main API class is *dqm4hep::core::Signal*. The following code illustrates how to use of this class with a simple sender/receive pattern:
+The DQM4hep core library provides a *dqm4hep::core::Signal* class that implements the observer pattern with an API similar to the Qt signal/slot mechanism. In contay to the Qt framework, the signal/slot mechanism described here is more likely a tool than a base framework functionality. The idea is to connect a function, or a set of functions, to process when a signal is emitted. These functions can be global functions or class methods. The following code illustrates how to use of the *dqm4hep::core::Signal* class with simple sender/receiver classes:
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: > The idea is to connect a function, or a set of functions, to process when a signal is emitted.

The sentence feels weird. Maybe something more like : 
The idea is to process a function, or set of functions, when a signal is emitted.

File: docs/user-manual/core-tools.md [hunk 3]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+
+## XML parsing features
+
+### XML file includes
+
+The element `<include>` allows to include an other XML file in-place of the element. The attribute `ref` of the XML element points on a XML using either an absolute or relative path. If a relative path is used, it is relative to the file containing the XML include element. By default, the XML also process nested include elements, meaning that *file1.xml* can include *file2.xml* that can also includes *file3.xml*.
+
+<div class="warning-msg">
+  <i class="fa fa-warning"></i>
+  No check is perform to protect against infinite recursive includes. It is users responsability to check the file consistency before using the XML file parser.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: No check is **performed**.

File: docs/user-manual/core-tools.md [hunk 4]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+</dqm4hep>
+```
+
+### MySQL database parameter select
+
+To be done - requires documentation on MySQL parameter database and tables ...
+
+### Use of environment variables
+
+Constants are a nice feature of the DQM4hep XML parser but can sometimes be a problem when users are dealing with passwords and other private sensitive data such API tokens. Environment variables are often use to deal with this problem as they are defined locally in your shell and thus not accessible you opening and reading the XML file in an editor. The XML parser can use environment variables by using a similar syntax as for constants (see above) with `$ENV{var}` where `var` is an environment variable.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: ... private sensitive data **such as** API tokens.

Environment variables are often **used** to deal with...

thus not accessible **to** you **when/while?** opening and reading the XML file in an editor.

File: docs/user-manual/core-tools.md [hunk 5]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+
+Example:
+
+```xml
+<dqm4hep>
+  <!-- Oops ! Password hardcoded... -->
+  <login user="Arthur" password="Cuillere"/>
```
Comment (apingault) [inline docs/user-manual/core-tools.md:396]: >Pas changer assiette pour fromache !

File: docs/user-manual/core-tools.md [hunk 6]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+  <!-- The current value of the counter is 1 -->
+  <counter value="1"/>
+  <!-- The current value of the counter is 2 -->
+  <counter value="2"/>
+  <!-- The current value of the counter is 3 -->
+  <counter value="3"/>
+</dqm4hep>
+```
+
+By construction, it is also possible the run nested for loops by changing the loop id:
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: By construction, it is also possible **to make** nested for loops by changing the loop id:

File: docs/user-manual/core-tools.md [hunk 7]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+  <entry x="1" y="0"/>
+  <entry x="1" y="1"/>
+  <entry x="1" y="2"/>
+  <entry x="2" y="0"/>
+  <entry x="2" y="1"/>
+  <entry x="2" y="2"/>
+</dqm4hep>
+```
+
+The `<for>` element attributes possible values are summurized in table below.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: ** Possible values for the `<for>` element attribute are summarized in the table below. **

File: docs/user-manual/core-tools.md [hunk 8]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+
+- Includes resolution
+- Constants and environment variables
+- Database connections and parameter select
+- For loops
+
+This has the following effects:
+
+1. Constants and environment variables can be used in database connections and for loops
+2. As includes are processed before constants, the `ref` attribute of the `<include>` element can not refer to any constant. In this case the constant will not replaced and the file will not been included.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: In this **case,** the constant will not **be** replaced and the file will not **be** included.

File: docs/user-manual/core-tools.md [hunk 9]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+3. As constants are processed after includes, it possible to define constants in an included file. The `<constants>` section in the included file has to be defined in the root element and the `<include>` element in the main file has to be defined in the root element.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: As constants are processed after includes, it **is** possible to define constants in an included file.

File: docs/user-manual/core-tools.md [hunk 10]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+}
+```
+
+The XML document itself can be accessed with the method `XMLParser::document()`:
+
+```cpp
+auto document = parser.document();
+```
+
+All of the features listed above are by default active but can be de-activated using dedicated method. The example below shows how to de-activate all the features listed above.
```
Comment (apingault) [inline docs/user-manual/core-tools.md]: All of the features listed above are by default active but can be de-activated using dedicated **methods**.
**A warning is sent to the logger when a feature is manually deactivated**.

**Review by apingault — COMMENTED** [review]
I have just quickly reviewed the `core-tools.md` for now. Will do the rest later

File: docs/user-manual/core-tools.md [hunk 11]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+  STATUS_CODE_ALREADY_PRESENT,
+  STATUS_CODE_OUT_OF_RANGE,
+  STATUS_CODE_NOT_ALLOWED,
+  STATUS_CODE_INVALID_PARAMETER,
+  STATUS_CODE_UNCHANGED,
+  STATUS_CODE_INVALID_PTR
+};
+```
+
+The *StatusCodeException* exception class handle one of these status codes on creation. Together with the StatusCode enum are also defined the following pre-processor macros:
```
Comment (tcoates3) [inline docs/user-manual/core-tools.md]: A little awkward phrasing, change to something like:

> The StatusCodeException exception class **handles** one of these status codes on creation. Together with the StatusCode enum, **there are also definitions for** the following pre-processor macros

File: docs/user-manual/core-tools.md [hunk 12]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+
+- *RETURN_RESULT_IF(code, Operator, Command)*: return the status code returned by the command if the comparison code with *code* using the operator *Operator* fails. See example below.
```
Comment (tcoates3) [inline docs/user-manual/core-tools.md]: The phrasing for both `RETURN_RESULT_IF` and `RETURN_RESULT_IF_AND_IF` are a little hard to follow. Maybe something more like:

> Runs the _command_, and returns it's status code if the comparison with _code_ using the _operator_ is false.
> Runs the _command_, and returns it's status code if the comparison with either _code1_ or _code2_ using the _operator_ is false.

File: docs/user-manual/core-tools.md [hunk 13]
```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
...
+myFunction(42) return STATUS_CODE_INVALID_PARAMETER
+    in function: doIt
+    in file:     /path/to/toto.cc line#: 75
+```
+
+These macros are particuliarly useful to debug a huge stack trace by following the function call by function name, file and line call.
 
+# The XML parser
+
+One of the nice components of the DQM4hep core library is the XML parser. It uses the [tinyxml](http://www.grinninglizard.com/tinyxml/) library to internally parse a XML document. On top of that comes a certain number of features that modifies the XML tree in memory after parsing listed below.
```
Comment (tcoates3) [inline docs/user-manual/core-tools.md]: A few phrasing/grammatical changes:

> [...] to internally parse **an** XML document. **In addition, it comes with a number of features for modifying the XML tree in memory after parsing, listed below.**

File: docs/user-manual/event-interface.md [hunk 14]
```diff
@@ -268,9 +233,124 @@ StatusCode DeviceEventStreamer::read(Event*& pEvent, xdrstream::IODevice* pDevic
...
+  for(int e=0 ; e<nevents ; e++) {
+    // hypotetical function to read an event from your file
+    MyEventType* event = this->performReaderEvent(); 
+  }
+  return STATUS_CODE_SUCCESS;
+}
+```  
+
+Very often, event data model libraries provides a mechanism to efficiently skip N events from a file without having to unpack events one by one as it done in the example above.
+The `runInfo()` instruct the reader to find the run information related to the file currently being read. If such an information is not available, it is still possible to provide dummy information.
```
Comment (tcoates3) [inline docs/user-manual/event-interface.md]: Just grammar again: 
> The `runInfo()` **instructs** the reader to find the run information related to the file currently being read. If **this information** is not available, it is still possible to provide dummy information.

**Comment by rete** [comment]
Thanks for the review. I will try to implement the corrections this afternoon.

File: mkdocs.yml (hunk 1 above)
Comment (rete) [reply mkdocs.yml:9]: OK

File: docs/user-manual/core-tools.md (hunk 2 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 3 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 4 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 6 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 5 above)
Comment (rete) [reply docs/user-manual/core-tools.md:396]: Arthouuuur ! La fleur en bouquet faaaaane, mais jeumais ne reunaaaaait !!

File: docs/user-manual/core-tools.md (hunk 7 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 8 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 11 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/core-tools.md (hunk 12 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: it's -> its, mister ! ^^

File: docs/user-manual/core-tools.md (hunk 13 above)
Comment (rete) [reply docs/user-manual/core-tools.md]: OK

File: docs/user-manual/event-interface.md (hunk 14 above)
Comment (rete) [reply docs/user-manual/event-interface.md]: OK

**Review by tcoates3 — DISMISSED** [review]

**Comment by rete** [comment]
I will not implement for now the MySQL feature documentation. It is quite long  and maybe not so important yet
